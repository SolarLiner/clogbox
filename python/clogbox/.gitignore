.DS_Store
__pycache__
/junit
.coverage
coverage.xml