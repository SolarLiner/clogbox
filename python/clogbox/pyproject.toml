[project]
name = "clogbox"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
authors = [
    { name = "Nathan Graule", email = "solarliner@gmail.com" }
]
requires-python = ">=3.11"
dependencies = [
    "sympy>=1.14.0",
]

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[dependency-groups]
dev = [
    "pytest>=8.3.5",
    "syrupy>=4.9.1"
]
