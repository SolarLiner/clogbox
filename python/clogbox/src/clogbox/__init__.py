def main() -> None:
    print("Hello from clogbox!")
