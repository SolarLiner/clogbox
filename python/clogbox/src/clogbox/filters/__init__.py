import typing

import sympy as sp


class IntegratorInput(typing.NamedTuple):
    s: sp.Basic
    g: sp.Basic
    x: sp.Basic

Integrator: typing.TypeAlias = typing.Callable[[IntegratorInput], sp.Basic]


def linear_integrator(inp: IntegratorInput) -> sp.Basic:
    return inp.g * inp.x + inp.s


def ota_integrator(drive: sp.Basic, tanh=sp.tanh) -> typing.Callable[[IntegratorInput], sp.Basic]:
    return lambda inp: inp.g * tanh(drive * inp.x) / drive + inp.s


Shaper: typing.TypeAlias = typing.Callable[[sp.Basic], sp.Basic]

def drive(shaper: Shaper, drive: sp.Basic) -> Shaper:
    return lambda x: shaper(drive * x) / drive


def bias(shaper: Shaper, bias: sp.Basic) -> Shaper:
    return lambda x: shaper(x + bias)
