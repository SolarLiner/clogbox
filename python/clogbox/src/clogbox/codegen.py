import io
import typing
from dataclasses import dataclass
from typing import Optional, Tuple, Union

import sympy as sp
from sympy.codegen.ast import value_const
from sympy.codegen import ast
from sympy.printing.codeprinter import requires
from sympy.printing.rust import RustCodePrinter
import sympy.matrices.expressions as mat
from sympy.utilities import codegen

from clogbox.nodes import Tuple

RUST_CODE_PREAMBLE = """
//! Generated source code with the `clogbox` Python package, to use with the `clogbox` project.
#![allow(unused_imports, dead_code, non_snake_case, non_camel_case)]
use num_traits::{Float, FloatConst};
use az::CastFrom;
"""

DIFFERENTIABLE_CODE_PREAMBLE = (
    RUST_CODE_PREAMBLE
    + """
use clogbox_math::root_eq;
use nalgebra as na;
"""
)


class ClogboxRustCodePrinter(RustCodePrinter):
    tab = " " * 4

    def __init__(self, settings=None):
        if settings is None:
            settings = {}
        user_functions = settings.pop("user_functions", {})
        user_functions.update({"Determinant": "determinant"})
        settings["user_functions"] = user_functions
        super().__init__(settings)

        self.uses: set[str] = set()
        self.type_params: set[str] = set()

    @requires(type_params={"Float"})
    def _print_Pow(self, expr):
        base_code = self._print(expr.base)
        exp = expr.exp
        if isinstance(exp, sp.Integer):
            return f"{base_code}.powi({int(exp)})"
        else:
            return super()._print_Pow(expr)

    @requires(type_params={"CastFrom<f64>"}, uses={"az::CastFrom"})
    def _print_Integer(self, expr, _type=False):
        return f"T::cast_from({float(expr)})"

    @requires(type_params={"CastFrom<f64>"}, uses={"az::CastFrom"})
    def _print_Float(self, expr, **kwargs):
        return f"T::cast_from({float(expr)})"

    def _print_Rational(self, expr):
        p, q = tuple(self._print(sp.Integer(x)) for x in (expr.p, expr.q))
        return f"{p}/{q}"

    @requires(type_params={"na::Scalar"}, uses={"nalgebra as na"})
    def _print_MatrixBase(self, mat: sp.MatrixBase):
        values = ", ".join(self._print(x) for x in mat)
        cdim = f"na::Const<{mat.cols}>"
        if mat.rows == 1:
            mtype = f"na::OVector::<T, {cdim}>"
        else:
            rdim = f"na::Const<{mat.rows}>"
            mtype = f"na::OMatrix::<T, {rdim}, {cdim}>"
        return f"{mtype}::new({values})"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_NaN(self, expr, _type=False):
        return "T::nan()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_Exp1(self, expr, _type=False):
        return "T::E()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_Pi(self, expr, _type=False):
        return "T::PI()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_Infinity(self, expr, _type=False):
        return "T::INFINITY()"

    @requires(type_params={"FloatConst"}, uses={"num_traits::FloatConst"})
    def _print_NegativeInfinity(self, expr, _type=False):
        return "T::NEG_INFINITY()"

    def _print_Tuple(self, expr: Tuple):
        return "({values})".format(values=", ".join(self._print(x) for x in expr.args))

    def _print_FunctionDefinition(self, fdecl: ast.FunctionDefinition):
        self.type_params.clear()

        params = ", ".join(self._print_Variable(var, _type=True) for var in fdecl.parameters)
        rtype = self._print_Type(fdecl.return_type)
        body = self._indent_codestring(self._print(fdecl.body))
        typeparams = "<T{bounds}>".format(bounds=": " + " + ".join(self.type_params) if len(self.type_params) > 0 else "")
        return f"""fn {fdecl.name}{typeparams}({params}) -> {rtype} {{\n{body}\n}}"""

    def _print_Type(self, t: ast.Type):
        if isinstance(t, ast.ComplexBaseType):
            self.uses.add("nalgebra as na")
            return "na::Complex<T>"
        if isinstance(t, ast.SignedIntType):
            return "i64"
        if isinstance(t, ast.UnsignedIntType):
            return "u64"
        return "T"

    def _print_Declaration(self, decl: ast.Declaration, _type=False):
        if value_const in decl.variable.attrs:
            mut = ""
        else:
            mut = " mut"
        return self._get_statement(f"let{mut} {self._print_Variable(decl.variable, _type=_type)} = {self._print(decl.variable.value)}")

    def _print_Variable(self, x: ast.Variable, _type=False):
        if _type:
            return f"{x.symbol.name}: {self._print_Type(x.type)}"
        return x.symbol.name

    def _print_While(self, node: ast.While):
        condition = self._print(node.condition)
        body = self._indent_codestring(self._print(node.body))
        return f"while {condition} {{\n{body}\n}}"

    def _print_Return(self, node: ast.Return, is_last_statement=False):
        if is_last_statement:
            return self._print(getattr(node, "return"))
        return self._get_statement(f"return {self._print(getattr(node, 'return'))}")

    def _print_Determinant(self, expr: mat.Determinant):
        return self._print_Function(expr)

    def _indent_codestring(self, codestring):
        return '\n'.join([self.tab + line for line in codestring.split('\n')])


class ClogboxCodegen(codegen.RustCodeGen):
    printer: ClogboxRustCodePrinter
    def __init__(self, **kwargs):
        kwargs["printer"] = ClogboxRustCodePrinter()
        super().__init__(**kwargs)

    def get_prototype(self, routine: codegen.Routine):
        def print_arg(arg: codegen.InputArgument | codegen.InOutArgument):
            self.printer.type_params.clear()

            typ = get_type(arg.name)
            if typ.startswith("na::"):
                self.printer.uses.add("nalgebra as na")
                self.printer.type_params.add("na::Scalar")
            return f"{arg.name}: {typ}"

        replacement, results = sp.cse(result.expr for result in typing.cast(list[codegen.Result], routine.results))
        codeblock = ast.CodeBlock(*(ast.Assignment(lhs, rhs) for lhs, rhs in replacement), ast.Return(*results))
        _body = self.printer.indent_code(self.printer.doprint(codeblock))
        args = ", ".join(print_arg(arg) for arg in routine.arguments)
        typeparams = "<T{bounds}>".format(bounds=": " + " + ".join(self.printer.type_params) if len(
            self.printer.type_params) > 0 else "")
        return f"fn {routine.name}{typeparams}({args}) -> T"


def render_as_module(content, printer: Optional[ClogboxRustCodePrinter]=None) -> str:
    if not printer:
        printer = ClogboxRustCodePrinter()
    str = printer.doprint(content)
    imports = "\n".join(f"use {module};" for module in printer.uses)
    return f"{imports}\n\n{str}"


def codegen_module(routines, printer: Optional[ClogboxRustCodePrinter]=None, codegen: Optional[ClogboxCodegen]=None,
                   project: str = "clogbox") -> str:
    if not printer:
        printer = ClogboxRustCodePrinter()
    if not codegen:
        codegen = ClogboxCodegen(printer=printer)

    [(_, contents)] = codegen.write(routines, "")  # type: str
    imports = "\n".join(f"use {module};" for module in codegen.printer.uses)
    end_of_comment = contents.find("*/") + 3 # 2 chars + newline
    return contents[:end_of_comment] + "\n" + imports + "\n\n" + contents[end_of_comment:]


def get_type(e: sp.Basic):
    """
    Get the Rust type of a sympy expression
    :param e: Sympy expression
    :return: Rust type representation
    """
    match e:
        case sp.MatrixBase(cols=c, rows=r) | sp.MatrixSymbol(cols=c, rows=r):
            return f"na::OMatrix<T, na::Const<{r}>, na::Const<{c}>>"
        case _:
            return "T"


@dataclass()
class IndentStream(io.TextIOBase):
    stream: io.TextIOBase
    indent: str = " " * 4

    def write(self, s: str) -> int:
        return self.stream.write(s.replace("\n", "\n" + self.indent))

    def flush(self):
        self.stream.flush()

    def close(self):
        self.stream.close()


def generate_codeblock(exprs: list[sp.Expr]) -> ast.CodeBlock:
    replacements, exprs = sp.cse(exprs)
    return ast.CodeBlock(*(ast.Assignment(lhs, rhs) for lhs, rhs in replacements), Tuple(exprs))

def generate_differentiable(
    f: io.TextIOBase,
    expr: Union[sp.Expr, sp.Eq],
    variable: Union[sp.Symbol, sp.MatrixBase],
    struct_name="Equation",
    printer: Optional[RustCodePrinter] = None,
    codegen: Optional[ClogboxCodegen] = None,
    runtime_invert=False,
):
    """
    Generate Rust code for a specific equation that is differentiable once about a variable. This function will
    implement the `clogbox_math::root_eq::Differentiable` trait for the generated struct.

    The resulting source code needs the `clogbox_math` and `num-traits` crates. The final module must also have the
    contents of the `DIFFERENTIABLE_CODE_PREAMBLE` constant present.

    :param f: Stream to write the code to.
    :param expr: Expression to implement the trait for.
    :param variable: Variable about which the equation is differentiable.
    :param struct_name: Name of the struct to generate (defaults to "Equation").
    :param printer: Optional instance of `RustCodePrinter` used to print symbolic
        expressions as Rust-compatible code. If not provided, a default printer will
        be used.
    """
    if not printer:
        printer = ClogboxRustCodePrinter()
    if not codegen:
        codegen = ClogboxCodegen(printer=printer)

    if isinstance(expr, sp.Eq):
        expr = expr.rhs - expr.lhs

    if isinstance(variable, sp.MatrixBase):
        var_symbols = set(variable.free_symbols)
    else:
        var_symbols = {variable}
    fields = sorted(expr.free_symbols - var_symbols, key=lambda s: s.name)

    f.write(f"pub struct {struct_name}<T> {{")
    for field in fields:
        f.write(f"    pub {str(field)}: {get_type(field)},")
    f.write( "}")
    f.writelines([""])

    if isinstance(expr, sp.MatrixBase):
        vtype = "na::OVector<Self::Scalar, Self::Dim>"
        mtype = "na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>"
        vview_type = "na::VectorView<Self::Scalar, Self::Dim>"
        f.write(
f"""
impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for {struct_name}<T> {{
    type Scalar = T;
    type Dim = na::Const<{expr.rows}>;
    
    fn eval_with_inv_jacobian(&self, matrix: {vview_type}) -> ({vtype}, {mtype}) {{"""
        )
        for i, name in enumerate(variable):
            f.write(f"        let {str(name)} = matrix[{i}];")
        for field in fields:
            f.write(f"        let {str(field)} = self.{str(field)};")

        s = IndentStream(f, indent=" " * 8)
        if runtime_invert:
            j = expr.jacobian(variable)
            block = generate_codeblock([expr, j])
            s.write(printer.doprint(block, assign_to="(f, df)"))
            s.write("        (f, df.try_inverse().unwrap())")
        else:
            inv_j = expr.jacobian(variable).inverse()
            block = generate_codeblock([expr, inv_j])
            s.write(printer.doprint(block))

    else:
        f.write(
f"""
impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::Differentiable for {struct_name}<T> {{
    type Scalar = T;

    fn eval_with_derivative(&self, {str(variable)}: T) -> (T, T) {{
"""
        )
        for field in fields:
            f.write(f"        let {str(field)} = self.{str(field)};")

        diff = sp.diff(expr, variable)
        block = generate_codeblock([expr, diff])
        f.write(printer.doprint(block))
        f.write(
"""
    }
}
"""
        )