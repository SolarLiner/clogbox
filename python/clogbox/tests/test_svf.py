from io import StringIO

import sympy as sp

import pytest

from clogbox.codegen import ClogboxRustCodePrinter
from clogbox.filters import Integrator, Shaper, linear_integrator, ota_integrator
from clogbox.filters.svf import SvfInput

@pytest.mark.parametrize("integrator", [linear_integrator, ota_integrator(1)])
@pytest.mark.parametrize("shaper", [lambda x: x, sp.asinh, sp.tanh])
def test_svf(integrator: Integrator, shaper: Shaper, printer: ClogboxRustCodePrinter, snapshot: str):
    g, q = sp.symbols("g q", positive=True)
    svf = SvfInput(g, q, integrator, shaper).generate()
    actual = StringIO()
    svf.generate_module(actual, printer=printer)
    assert snapshot == actual.getvalue()
