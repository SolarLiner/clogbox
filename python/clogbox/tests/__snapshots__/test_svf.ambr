# serializer version: 1
# name: test_svf[<lambda>-<lambda>]
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use az::CastFrom;
  use clogbox_math::root_eq;
  use nalgebra as na;
  use num_traits::Float;
  use num_traits::FloatConst;
  
  
  pub fn state<T: CastFrom<f64> + Float + na::Scalar>(S: na::OVector<T, na::Const<2>>, g: T, y_bp: T, y_hp: T, y_lp: T) -> na::OVector<T, na::Const<2>> {
  
      let S = na::OVector::<T, na::Const<2>>::new(g*y_hp.tanh() + y_bp, g*y_bp.tanh() + y_lp);
      S
  
  }
  
  
  pub struct SvfEquation<T: na::Scalar> {
      pub S: na::OVector<T, na::Const<2>>,
      pub g: T,
      pub q: T,
      pub x: T,
  }
  
  impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for SvfEquation<T> {
      type Scalar = T;
      type Dim = na::Const<3>;
      
      fn eval_with_inv_jacobian(&self, matrix: na::VectorView<Self::Scalar, Self::Dim>) -> (na::OVector<Self::Scalar, Self::Dim>, na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>) {
          let y_lp = matrix[0];
          let y_bp = matrix[1];
          let y_hp = matrix[2];
          let S = self.S;
          let g = self.g;
          let q = self.q;
          let x = self.x;
          let (f, df) = {
              let x0 = T::cast_from(4.0)*q;
              let x1 = -T::cast_from(2.0)*q*y_bp - T::cast_from(0.5)*x + y_bp + T::cast_from(0.5)*y_hp + T::cast_from(0.5)*y_lp;
              let x2 = T::cast_from(2.0)*q;
              let x3 = T::cast_from(1.0) - x2;
              let x4 = T::cast_from(2.0)*x3;
              let x5 = y_hp.tanh();
              let x6 = T::cast_from(2.0)*q;
              let x7 = (-T::cast_from(1.0)/T::cast_from(2.0)*x - x6*y_bp + y_bp + (T::cast_from(1.0)/T::cast_from(2.0))*y_hp + (T::cast_from(1.0)/T::cast_from(2.0))*y_lp).tanh();
              let x8 = x2 + T::cast_from(-1.0);
              let x9 = x7.powi(2);
              let x10 = g*(T::cast_from(1.0)/T::cast_from(2.0) - T::cast_from(1.0)/T::cast_from(2.0)*x9);
              (na::OVector::<T, na::Const<3>>::new(-x0*y_bp + x1*x4 + T::cast_from(2.0)*y_bp, g*x5 + x1 + S[0], -g*x7 - y_lp + S[1]), na::OMatrix::<T, na::Const<3>, na::Const<3>>::new(x3, -x0 - x4*x8 + T::cast_from(2.0), x3, T::cast_from(0.5), -x8, g*(T::cast_from(1.0) - x5.powi(2)) + T::cast_from(0.5), -x10 + T::cast_from(-1.0), -g*(T::cast_from(1.0) - x6)*(T::cast_from(1.0) - x9), -x10))
          };
          (f, df.try_inverse().unwrap())
              }
  }
  
  '''
# ---
# name: test_svf[<lambda>-linear_integrator]
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use az::CastFrom;
  use clogbox_math::root_eq;
  use nalgebra as na;
  use num_traits::Float;
  use num_traits::FloatConst;
  
  
  pub fn state<T: CastFrom<f64> + Float + na::Scalar>(S: na::OVector<T, na::Const<2>>, g: T, y_bp: T, y_hp: T, y_lp: T) -> na::OVector<T, na::Const<2>> {
  
      let S = na::OVector::<T, na::Const<2>>::new(g*y_hp + y_bp, g*y_bp + y_lp);
      S
  
  }
  
  
  pub struct SvfEquation<T: na::Scalar> {
      pub S: na::OVector<T, na::Const<2>>,
      pub g: T,
      pub q: T,
      pub x: T,
  }
  
  impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for SvfEquation<T> {
      type Scalar = T;
      type Dim = na::Const<3>;
      
      fn eval_with_inv_jacobian(&self, matrix: na::VectorView<Self::Scalar, Self::Dim>) -> (na::OVector<Self::Scalar, Self::Dim>, na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>) {
          let y_lp = matrix[0];
          let y_bp = matrix[1];
          let y_hp = matrix[2];
          let S = self.S;
          let g = self.g;
          let q = self.q;
          let x = self.x;
          let (f, df) = {
              let x0 = T::cast_from(4.0)*q;
              let x1 = -T::cast_from(2.0)*q*y_bp - T::cast_from(0.5)*x + y_bp + T::cast_from(0.5)*y_hp + T::cast_from(0.5)*y_lp;
              let x2 = -x1;
              let x3 = T::cast_from(2.0)*q;
              let x4 = T::cast_from(1.0) - x3;
              let x5 = T::cast_from(2.0)*x4;
              let x6 = x3 + T::cast_from(-1.0);
              let x7 = T::cast_from(0.5)*g;
              (na::OVector::<T, na::Const<3>>::new(-x0*y_bp - x2*x5 + T::cast_from(2.0)*y_bp, g*y_hp + x1 + S[0], g*x2 - y_lp + S[1]), na::OMatrix::<T, na::Const<3>, na::Const<3>>::new(x4, -x0 - x5*x6 + T::cast_from(2.0), x4, T::cast_from(0.5), -x6, g + T::cast_from(0.5), -x7 + T::cast_from(-1.0), g*x6, -x7))
          };
          (f, df.try_inverse().unwrap())
              }
  }
  
  '''
# ---
# name: test_svf[asinh-<lambda>]
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use az::CastFrom;
  use clogbox_math::root_eq;
  use nalgebra as na;
  use num_traits::Float;
  use num_traits::FloatConst;
  
  
  pub fn state<T: CastFrom<f64> + Float + na::Scalar>(S: na::OVector<T, na::Const<2>>, g: T, y_bp: T, y_hp: T, y_lp: T) -> na::OVector<T, na::Const<2>> {
  
      let S = na::OVector::<T, na::Const<2>>::new(g*y_hp.tanh() + y_bp, g*y_bp.tanh() + y_lp);
      S
  
  }
  
  
  pub struct SvfEquation<T: na::Scalar> {
      pub S: na::OVector<T, na::Const<2>>,
      pub g: T,
      pub q: T,
      pub x: T,
  }
  
  impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for SvfEquation<T> {
      type Scalar = T;
      type Dim = na::Const<3>;
      
      fn eval_with_inv_jacobian(&self, matrix: na::VectorView<Self::Scalar, Self::Dim>) -> (na::OVector<Self::Scalar, Self::Dim>, na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>) {
          let y_lp = matrix[0];
          let y_bp = matrix[1];
          let y_hp = matrix[2];
          let S = self.S;
          let g = self.g;
          let q = self.q;
          let x = self.x;
          let (f, df) = {
              let x0 = T::cast_from(2.0)*y_bp;
              let x1 = T::cast_from(4.0)*q;
              let x2 = T::cast_from(2.0)*q;
              let x3 = -T::cast_from(1.0)/T::cast_from(2.0)*x - x2*y_bp + y_bp + (T::cast_from(1.0)/T::cast_from(2.0))*y_hp + (T::cast_from(1.0)/T::cast_from(2.0))*y_lp;
              let x4 = x3.asinh();
              let x5 = T::cast_from(2.0)*q;
              let x6 = T::cast_from(1.0) - x5;
              let x7 = T::cast_from(2.0)*x6;
              let x8 = y_hp.tanh();
              let x9 = -q*x0 - T::cast_from(0.5)*x + y_bp + T::cast_from(0.5)*y_hp + T::cast_from(0.5)*y_lp;
              let x10 = x9.powi(2);
              let x11 = x10 + T::cast_from(1.0);
              let x12 = g*x11.sqrt().recip();
              let x13 = (x3.powi(2) + T::cast_from(1.0)).sqrt().recip();
              let x14 = T::cast_from(1.0)*x13*x6;
              let x15 = x13*(T::cast_from(1.0) - x2);
              let x16 = (T::cast_from(1.0)/T::cast_from(2.0))*x13;
              let x17 = x11.powf(T::cast_from(-1.5));
              let x18 = g*x17*x9*(T::cast_from(1.0)*q*y_bp + T::cast_from(0.25)*x - T::cast_from(0.5)*y_bp - T::cast_from(0.25)*y_hp - T::cast_from(0.25)*y_lp) + T::cast_from(0.5)*x12;
              (na::OVector::<T, na::Const<3>>::new(x0 - x1*y_bp + x4*x7, g*x8 + x4 + S[0], -x12*x9 - y_lp + S[1]), na::OMatrix::<T, na::Const<3>, na::Const<3>>::new(x14, -x1 + x15*x7 + T::cast_from(2.0), x14, x16, x15, g*(T::cast_from(1.0) - x8.powi(2)) + x16, -x18 + T::cast_from(-1.0), T::cast_from(0.5)*g*x10*x17*(T::cast_from(2.0) - x1) - x12*(T::cast_from(1.0) - x5), -x18))
          };
          (f, df.try_inverse().unwrap())
              }
  }
  
  '''
# ---
# name: test_svf[asinh-linear_integrator]
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use az::CastFrom;
  use clogbox_math::root_eq;
  use nalgebra as na;
  use num_traits::Float;
  use num_traits::FloatConst;
  
  
  pub fn state<T: CastFrom<f64> + Float + na::Scalar>(S: na::OVector<T, na::Const<2>>, g: T, y_bp: T, y_hp: T, y_lp: T) -> na::OVector<T, na::Const<2>> {
  
      let S = na::OVector::<T, na::Const<2>>::new(g*y_hp + y_bp, g*y_bp + y_lp);
      S
  
  }
  
  
  pub struct SvfEquation<T: na::Scalar> {
      pub S: na::OVector<T, na::Const<2>>,
      pub g: T,
      pub q: T,
      pub x: T,
  }
  
  impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for SvfEquation<T> {
      type Scalar = T;
      type Dim = na::Const<3>;
      
      fn eval_with_inv_jacobian(&self, matrix: na::VectorView<Self::Scalar, Self::Dim>) -> (na::OVector<Self::Scalar, Self::Dim>, na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>) {
          let y_lp = matrix[0];
          let y_bp = matrix[1];
          let y_hp = matrix[2];
          let S = self.S;
          let g = self.g;
          let q = self.q;
          let x = self.x;
          let (f, df) = {
              let x0 = T::cast_from(4.0)*q;
              let x1 = T::cast_from(2.0)*q;
              let x2 = -T::cast_from(1.0)/T::cast_from(2.0)*x - x1*y_bp + y_bp + (T::cast_from(1.0)/T::cast_from(2.0))*y_hp + (T::cast_from(1.0)/T::cast_from(2.0))*y_lp;
              let x3 = x2.asinh();
              let x4 = T::cast_from(1.0) - T::cast_from(2.0)*q;
              let x5 = T::cast_from(2.0)*x4;
              let x6 = (x2.powi(2) + T::cast_from(1.0)).sqrt().recip();
              let x7 = T::cast_from(1.0)*x4*x6;
              let x8 = x6*(T::cast_from(1.0) - x1);
              let x9 = (T::cast_from(1.0)/T::cast_from(2.0))*x6;
              let x10 = g*x9;
              (na::OVector::<T, na::Const<3>>::new(-x0*y_bp + x3*x5 + T::cast_from(2.0)*y_bp, g*y_hp + x3 + S[0], -g*x3 - y_lp + S[1]), na::OMatrix::<T, na::Const<3>, na::Const<3>>::new(x7, -x0 + x5*x8 + T::cast_from(2.0), x7, x9, x8, g + x9, -x10 + T::cast_from(-1.0), -g*x8, -x10))
          };
          (f, df.try_inverse().unwrap())
              }
  }
  
  '''
# ---
# name: test_svf[tanh-<lambda>]
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use az::CastFrom;
  use clogbox_math::root_eq;
  use nalgebra as na;
  use num_traits::Float;
  use num_traits::FloatConst;
  
  
  pub fn state<T: CastFrom<f64> + Float + na::Scalar>(S: na::OVector<T, na::Const<2>>, g: T, y_bp: T, y_hp: T, y_lp: T) -> na::OVector<T, na::Const<2>> {
  
      let S = na::OVector::<T, na::Const<2>>::new(g*y_hp.tanh() + y_bp, g*y_bp.tanh() + y_lp);
      S
  
  }
  
  
  pub struct SvfEquation<T: na::Scalar> {
      pub S: na::OVector<T, na::Const<2>>,
      pub g: T,
      pub q: T,
      pub x: T,
  }
  
  impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for SvfEquation<T> {
      type Scalar = T;
      type Dim = na::Const<3>;
      
      fn eval_with_inv_jacobian(&self, matrix: na::VectorView<Self::Scalar, Self::Dim>) -> (na::OVector<Self::Scalar, Self::Dim>, na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>) {
          let y_lp = matrix[0];
          let y_bp = matrix[1];
          let y_hp = matrix[2];
          let S = self.S;
          let g = self.g;
          let q = self.q;
          let x = self.x;
          let (f, df) = {
              let x0 = T::cast_from(4.0)*q;
              let x1 = T::cast_from(2.0)*q;
              let x2 = (-T::cast_from(1.0)/T::cast_from(2.0)*x - x1*y_bp + y_bp + (T::cast_from(1.0)/T::cast_from(2.0))*y_hp + (T::cast_from(1.0)/T::cast_from(2.0))*y_lp).tanh();
              let x3 = T::cast_from(2.0) - T::cast_from(4.0)*q;
              let x4 = y_hp.tanh();
              let x5 = x2.tanh();
              let x6 = x2.powi(2);
              let x7 = (T::cast_from(1.0)/T::cast_from(2.0))*x6;
              let x8 = T::cast_from(1.0)/T::cast_from(2.0) - x7;
              let x9 = x3*x8;
              let x10 = (T::cast_from(1.0) - x1)*(T::cast_from(1.0) - x6);
              let x11 = g*(T::cast_from(1.0) - x5.powi(2));
              let x12 = x11*x8;
              (na::OVector::<T, na::Const<3>>::new(-x0*y_bp + x2*x3 + T::cast_from(2.0)*y_bp, g*x4 + x2 + S[0], -g*x5 - y_lp + S[1]), na::OMatrix::<T, na::Const<3>, na::Const<3>>::new(x9, -x0 + x10*x3 + T::cast_from(2.0), x9, x8, x10, g*(T::cast_from(1.0) - x4.powi(2)) - x7 + T::cast_from(1.0)/T::cast_from(2.0), -x12 + T::cast_from(-1.0), -x10*x11, -x12))
          };
          (f, df.try_inverse().unwrap())
              }
  }
  
  '''
# ---
# name: test_svf[tanh-linear_integrator]
  '''
  /*
   *                      Code generated with SymPy 1.14.0
   *
   *              See http://www.sympy.org/ for more information.
   *
   *                       This file is part of 'project'
   */
  
  #![allow(unused_imports, dead_code, non_snake_case, non_camel_case_types)]
  
  use az::CastFrom;
  use clogbox_math::root_eq;
  use nalgebra as na;
  use num_traits::Float;
  use num_traits::FloatConst;
  
  
  pub fn state<T: CastFrom<f64> + Float + na::Scalar>(S: na::OVector<T, na::Const<2>>, g: T, y_bp: T, y_hp: T, y_lp: T) -> na::OVector<T, na::Const<2>> {
  
      let S = na::OVector::<T, na::Const<2>>::new(g*y_hp + y_bp, g*y_bp + y_lp);
      S
  
  }
  
  
  pub struct SvfEquation<T: na::Scalar> {
      pub S: na::OVector<T, na::Const<2>>,
      pub g: T,
      pub q: T,
      pub x: T,
  }
  
  impl<T: Copy + na::Scalar + na::RealField + FloatConst + CastFrom<f64>> root_eq::MultiDifferentiable for SvfEquation<T> {
      type Scalar = T;
      type Dim = na::Const<3>;
      
      fn eval_with_inv_jacobian(&self, matrix: na::VectorView<Self::Scalar, Self::Dim>) -> (na::OVector<Self::Scalar, Self::Dim>, na::OMatrix<Self::Scalar, Self::Dim, Self::Dim>) {
          let y_lp = matrix[0];
          let y_bp = matrix[1];
          let y_hp = matrix[2];
          let S = self.S;
          let g = self.g;
          let q = self.q;
          let x = self.x;
          let (f, df) = {
              let x0 = T::cast_from(4.0)*q;
              let x1 = T::cast_from(2.0)*q;
              let x2 = (-T::cast_from(1.0)/T::cast_from(2.0)*x - x1*y_bp + y_bp + (T::cast_from(1.0)/T::cast_from(2.0))*y_hp + (T::cast_from(1.0)/T::cast_from(2.0))*y_lp).tanh();
              let x3 = T::cast_from(2.0) - T::cast_from(4.0)*q;
              let x4 = x2.powi(2);
              let x5 = (T::cast_from(1.0)/T::cast_from(2.0))*x4;
              let x6 = T::cast_from(1.0)/T::cast_from(2.0) - x5;
              let x7 = x3*x6;
              let x8 = (T::cast_from(1.0) - x1)*(T::cast_from(1.0) - x4);
              let x9 = g*x6;
              (na::OVector::<T, na::Const<3>>::new(-x0*y_bp + x2*x3 + T::cast_from(2.0)*y_bp, g*y_hp + x2 + S[0], -g*x2 - y_lp + S[1]), na::OMatrix::<T, na::Const<3>, na::Const<3>>::new(x7, -x0 + x3*x8 + T::cast_from(2.0), x7, x6, x8, g - x5 + T::cast_from(1.0)/T::cast_from(2.0), -x9 + T::cast_from(-1.0), -g*x8, -x9))
          };
          (f, df.try_inverse().unwrap())
              }
  }
  
  '''
# ---
